sdk/database/newdbplugin/database.go

```diff
@@ -0,0 +1,155 @@
+// map[string]interface{}. It does have an `Any` type, but converting it to a map
+// requires extensive use of reflection and knowing what types to support ahead of
+// time. Instead these types are made as user-friendly as possible so the conversion
+// between protobuf types and request/response objects is handled by Vault developers
+// rather than needing to be handled by external plugin developers.
+// ///////////////////////////////////////////////////////////////////////////
+
+// ///////////////////////////////////////////////////////
+// Initialize()
+// ///////////////////////////////////////////////////////
+
+// InitializeRequest contains all information needed to initialize a database plugin.
+type InitializeRequest struct {
+	// Config to initialize the database with. This can include things like connection details,
+	// a "root" username & password, etc. This will not include all configuration items specified
+	// when configuring the database. Some values will be stripped out by the database engine
+	// prior to being passed to the plugin.
+	Config map[string]interface{}
+
+	// VerifyConnection during initialization. If true, a connection should be made to the
+	// database to verify the connection can be made. If false, no connection should be made
+	// on initialization.
+	VerifyConnection bool
+}
+
+// InitializeResponse returns any information Vault needs to know after initializing
+// a database plugin.
+type InitializeResponse struct {
+	// Config that should be saved in Vault. This may differ from the config in the request,
+	// but should contain everything required to Initialize the database.
+	Config map[string]interface{}
+}
+
+// ///////////////////////////////////////////////////////
+// NewUser()
+// ///////////////////////////////////////////////////////
+
+// NewUserRequest request a new user is created
+type NewUserRequest struct {
+	// UsernameConfig is metadata that can be used to generate a username
+	// within the database plugin
+	UsernameConfig UsernameMetadata
+
+	// Statements is an ordered list of commands to run within the database when creating a new user.
+	// This frequently includes permissions to give the user or similar actions.
+	Statements Statements
+
+	// Password credentials to use when creating the user
+	Password string
+
+	// Expiration of the user. Not all database plugins will support this.
+	Expiration time.Time
+}
+
+// UsernameMetadata is metadata the database plugin can use to generate a username
+type UsernameMetadata struct {
+	DisplayName string
+	RoleName    string
+}
+
+// NewUserResponse returns any information Vault needs to know after creating a new user.
+type NewUserResponse struct {
+	// Username of the user created within the database.
+	// REQUIRED
+	Username string
+}
+
+// ///////////////////////////////////////////////////////
+// UpdateUser()
+// ///////////////////////////////////////////////////////
+
+type UpdateUserRequest struct {
```

Comment:
These are the statements that are currently supported and how they should be mapped to this interface:
`root_rotation_statements` -> `UpdateUserRequest.ChangePassword.Statements` (missing currently - will add)
`creation_statements` -> `NewUserRequest.Statements`
`revocation_statements` -> `DeleteUserRequest.Statements` (missing currently - will add)
`rollback_statements` -> `NewUserRequest.RollbackStatements` (missing currently - will add) - I hate this one because the description is specific to user creation but the name is general
`renew_statements` -> `UpdateUserRequest.ChangeExpiration.Statements`
`rotation_statements` -> `UpdateUserRequest.ChangePassword.Statements`

Reply:
It looks like you've added statements to `ChangePassword` and `ChangeExpiration`. I am not concerned that they are not in the parent struct as long as they are accessible through those fields.